File: my-solutions/0098-validate-binary-search-tree/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:

    bool isValid(TreeNode* root, TreeNode*& prev) {
        // check if the root is nullptr;
        if(root == nullptr) return true;

        if(!isValid(root->left, prev)) return false;

        if(prev && prev->val >= root->val) return false;
        prev =  root;

        if(!isValid(root->right, prev)) return false;

        return true;
    }

    bool isValidBST(TreeNode* root) {
        TreeNode* prev = nullptr;
        return isValid(root, prev);
    }
};

File: my-solutions/0098-validate-binary-search-tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:

        res = True

        # maxLeft, minLeft, maxRight, minRight
        def dfs(node):
            nonlocal res
            if not node:
                return float('inf'), float('-inf')

            lmin, lmax = dfs(node.left)
            rmin, rmax = dfs(node.right)

            if lmax >= node.val or rmin <= node.val:
                res = False

            return min([node.val, lmin, rmin]), max([node.val, lmax, rmax])

        dfs(root)

        return res


